static/src/javascripts/projects/commercial/modules/dfp/Advert.ts

```diff
@@ -0,0 +1,150 @@
+const createSizeMapping = (attr: string): AdSize[] =>
+	attr
+		.split('|')
+		.map((size) =>
+			size === 'fluid' ? 'fluid' : size.split(',').map(Number),
+		);
+
+const getAdBreakpointSizes = (advertNode: HTMLElement): AdSizes =>
+	breakpoints.reduce<Record<string, AdSize[]>>((sizes, breakpoint) => {
+		const data = advertNode.getAttribute(
+			`data-${breakpointNameToAttribute(breakpoint.name)}`,
+		);
+		if (data) {
+			sizes[breakpoint.name] = createSizeMapping(data);
+		}
+		return sizes;
+	}, {});
+
+class Advert {
+	id: string;
+	node: HTMLElement;
+	sizes: AdSizes;
+	size: AdSize | null = null;
+	slot: Record<string, unknown>;
+	isEmpty: boolean | null = null;
+	isLoading = false;
+	isRendering = false;
+	isLoaded = false;
+	isRendered = false;
+	shouldRefresh = false;
+	maxViewPercentage = 0;
+	whenLoaded: Promise<boolean>;
+	whenLoadedResolver: Resolver | null = null;
+	whenRendered: Promise<boolean>;
+	whenRenderedResolver: Resolver | null = null;
+	whenSlotReady: Promise<void>;
+	extraNodeClasses: string[] = [];
+	timings: Timings = {
+		createTime: null,
+		startLoading: null,
+		stopLoading: null,
+		startRendering: null,
+		stopRendering: null,
+		loadingMethod: null,
+		lazyWaitComplete: null,
+	};
+	hasPrebidSize = false;
+
+	constructor(adSlotNode: HTMLElement) {
+		const sizes = getAdBreakpointSizes(adSlotNode);
+		const slotDefinition = defineSlot(adSlotNode, sizes) as SlotDefinition;
+
+		this.id = adSlotNode.id;
+		this.node = adSlotNode;
+		this.sizes = sizes;
+		this.slot = slotDefinition.slot;
+
+		this.whenSlotReady = slotDefinition.slotReady;
+
+		this.whenLoaded = new Promise((resolve: Resolver) => {
+			this.whenLoadedResolver = resolve;
+		}).then((isLoaded: boolean): boolean => {
+			this.isLoaded = isLoaded;
+			return isLoaded;
+		});
+
+		this.whenRendered = new Promise((resolve: Resolver) => {
+			this.whenRenderedResolver = resolve;
+		}).then((isRendered) => {
+			this.isRendered = isRendered;
+			return isRendered;
+		});
+	}
+
+	static filterClasses = (
+		oldClasses: string[],
+		newClasses: string[],
+	): string[] =>
+		oldClasses.filter((oldClass) => !newClasses.includes(oldClass));
+
+	startLoading(): void {
+		this.isLoading = true;
+		this.timings.startLoading = getCurrentTime();
+	}
+
+	stopLoading(isLoaded: boolean): void {
+		this.isLoading = false;
+		if (this.whenLoadedResolver) {
+			this.whenLoadedResolver(isLoaded);
+		}
+		this.timings.stopLoading = getCurrentTime();
+	}
+
+	startRendering(): void {
+		this.isRendering = true;
+		this.timings.startRendering = getCurrentTime();
+	}
+
+	stopRendering(isRendered: boolean): void {
+		this.isRendering = false;
+		if (this.whenRenderedResolver) {
+			this.whenRenderedResolver(isRendered);
+		}
+	}
+
+	updateExtraSlotClasses(...newClasses: string[]): void {
+		const classesToRemove = Advert.filterClasses(
+			this.extraNodeClasses,
+			newClasses,
+		);
+		this.node.classList.remove(...classesToRemove);
+		this.node.classList.add(...newClasses);
```

Comment:
Sorry I didn't spot this earlier but I believe multiple args to classList is not supported by IE:
https://developer.mozilla.org/en-US/docs/Web/API/Element/classList#browser_compatibility

Reply:
This code is actually unchanged from the old Advert.js file, so if it's broken in IE it's been like that for a while. I can create a mini new PR to refactor into a loop though

Reply:
Oh sorry I assumed it was a conversion from bonzo `.addClass()`.
I guess we should be good boy scouts and fix it - don't worry I can do pick this up.

Reply:
https://github.com/guardian/frontend/pull/23820